Build failure with clang-3.4.1

- Neovim version:0.1.4
- Operating system/version: FreeBSD 10
- Terminal name/version:
- `$TERM`:

fails to build with clang-3.4.1:

```
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:32: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
static inline uint64_t be64toh(uint64_t big_endian_64_bits)
                       ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/endian.h:77:29: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                                 ^
/usr/include/sys/endian.h:61:30: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                                  ^
/usr/include/x86/endian.h:80:24: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
                              ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:24: error: expected ')'
static inline uint64_t be64toh(uint64_t big_endian_64_bits)
                       ^
/usr/include/sys/endian.h:77:20: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                        ^
/usr/include/sys/endian.h:61:20: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                        ^
/usr/include/x86/endian.h:80:27: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
                                 ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:24: note: to match this '('
/usr/include/sys/endian.h:77:20: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                        ^
/usr/include/sys/endian.h:61:20: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                        ^
/usr/include/x86/endian.h:80:2: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
        ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:24: error: static declaration of '__builtin_constant_p' follows non-static declaration
static inline uint64_t be64toh(uint64_t big_endian_64_bits)
                       ^
/usr/include/sys/endian.h:77:20: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                        ^
/usr/include/sys/endian.h:61:20: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                        ^
/usr/include/x86/endian.h:80:3: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
         ^
/usr/local/include/msgpack/pack_template.h:280:20: note: previous implicit declaration is here
    buf[0] = 0xcd; _msgpack_store16(&buf[1], d);
                   ^
/usr/local/include/msgpack/sysdep.h:156:25: note: expanded from macro '_msgpack_store16'
    do { uint16_t val = _msgpack_be16(num); memcpy(to, &val, 2); } while(0)
                        ^
/usr/local/include/msgpack/sysdep.h:94:34: note: expanded from macro '_msgpack_be16'
#        define _msgpack_be16(x) ntohs(x)
                                 ^
/usr/include/arpa/inet.h:173:18: note: expanded from macro 'ntohs'
#define ntohs(x)        __ntohs(x)
                        ^
/usr/include/x86/endian.h:129:20: note: expanded from macro '__ntohs'
#define __ntohs(x)      __bswap16(x)
                        ^
/usr/include/x86/endian.h:74:16: note: expanded from macro '__bswap16'
        ((__uint16_t)(__builtin_constant_p(x) ? \
                      ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:24: error: definition of builtin function '__builtin_constant_p'
static inline uint64_t be64toh(uint64_t big_endian_64_bits)
                       ^
/usr/include/sys/endian.h:77:20: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                        ^
/usr/include/sys/endian.h:61:20: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                        ^
/usr/include/x86/endian.h:80:3: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
         ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3232:32: error: parameter name omitted
static inline uint64_t be64toh(uint64_t big_endian_64_bits)
                               ^
/usr/include/sys/endian.h:77:29: note: expanded from macro 'be64toh'
#define be64toh(x)      bswap64((x))
                                 ^
/usr/include/sys/endian.h:61:30: note: expanded from macro 'bswap64'
#define bswap64(x)      __bswap64(x)
                                  ^
/usr/include/x86/endian.h:80:24: note: expanded from macro '__bswap64'
        (__builtin_constant_p(x) ?      \
                              ^
/usr/ports/editors/neovim/work/neovim-0.1.4/src/nvim/shada.c:3240:31: error: use of undeclared identifier 'big_endian_64_bits'
  uint8_t *buf = (uint8_t *) &big_endian_64_bits;
                              ^
1 warning and 5 errors generated.
ninja: build stopped: subcommand failed.
===> Compilation failed unexpectedly.
Try to set MAKE_JOBS_UNSAFE=yes and rebuild before reporting the failure to
the maintainer.
*** Error code 1

Stop.
make: stopped in /usr/ports/editors/neovim
```

--- a/config/CMakeLists.txt
+++ b/config/CMakeLists.txt
@@ -67,6 +67,14 @@ if(HAVE_LANGINFO_H)
   check_symbol_exists(CODESET "langinfo.h" HAVE_NL_LANGINFO_CODESET)
 endif()
 
+check_include_files("endian.h" HAVE_ENDIAN_H)
+check_include_files("sys/endian.h" HAVE_SYS_ENDIAN_H)
+
+set(ENDIAN_INCLUDE_FILE "endian.h")
+if(HAVE_SYS_ENDIAN_H AND NOT HAVE_ENDIAN_H)
+  set(ENDIAN_INCLUDE_FILE "sys/endian.h")
+endif()
+
 set(SI "#include <stdint.h>\n")
 set(MS "int main(int argc,char**argv)\n{\n  uint64_t i=0x0102030405060708ULL;")
 set(ME "}")
@@ -74,7 +82,7 @@ check_c_source_compiles("
   #define _BSD_SOURCE 1
   #define _DEFAULT_SOURCE 1
   ${SI}
-  #include <endian.h>
+  #include <${ENDIAN_INCLUDE_FILE}>
   #ifndef be64toh
   # error No be64toh macros
   #endif

--- a/config/config.h.in
+++ b/config/config.h.in
@@ -57,5 +57,6 @@
 
 #cmakedefine HAVE_BE64TOH
 #cmakedefine ORDER_BIG_ENDIAN
+#define ENDIAN_INCLUDE_FILE <@ENDIAN_INCLUDE_FILE@>
 
 #endif  // AUTO_CONFIG_H

--- a/src/nvim/shada.c
+++ b/src/nvim/shada.c
@@ -46,7 +46,7 @@
 #ifdef HAVE_BE64TOH
 # define _BSD_SOURCE 1
 # define _DEFAULT_SOURCE 1
-# include <endian.h>
+# include ENDIAN_INCLUDE_FILE
 #endif
 
 // Note: when using bufset hash pointers are intentionally casted to uintptr_t

